Added custom validator in user schema to ensure uniqueness of email

ashwaryy opened the pull request:
Hi. I have added a custom validator to check for unique email in user schema. There were two approaches to this:

1. Using `unique` option in mongoose: This approach would basically create a MongoDB index over email field and check the property for uniqueness. But the shortcoming of this approach is that the index must be created on the database for this to work. If index is not created, unique would not work. This is pointed out [here](https://mongoosejs.com/docs/validation.html#the-unique-option-is-not-a-validator).
2. Custom validator: Write a custom validator which searches the DB for documents using the email field and, if found, returns false. This is the approach I have used. 

Also, currently there is no pattern to match whether the string entered is actually an email. I think there should be a check on the backend to ensure whether the string entered is actually an email address, and the user has the ownership of that email. These two things can be accomplished by sending a code to the user's email and validating it on the backend. This way, validity of email and its ownership can be established. 

As of now, I have added a simple regex pattern to check whether the entered string is actually an email. But there has already been a lot of debate around this and the consensus is that regex is not the best way to do it. 

Maybe you can include email validation in future scope of work.

lugenx commented:
yeah, we need some more validations in the project. i am planning to install validator.js package and handle all of them with it, i would like to see your opinion about it too. i think regex can stay for now, thank you @ashwaryy , i will request other team members to review this before the merge.

ashwaryy commented:
There are two things here:

1. Validating that the string is an email
2. Verifying that the user has ownership of the email.

For the first one, validator package works. For the second, one way is to send a code to user's email and asking user to provide the same to backend. 

If ownership of email is not a concern right now, then validator would work.

AbhiBajpaishek commented on server/src/models/userModel.js:
```diff
@@ -10,7 +10,22 @@ const userSchema = new Schema(
       type: String,
-      required: true,
+      lowercase: true,
+      match: [/(.+)@(.+){2,}\.(.+){2,}/, "Please enter a valid email"],
```
Can we use the validators package?

ashwaryy replied:
If there's a plan to use it at multiple places or in all the models, then it makes sense to add it as a dependency and use it.

lugenx replied:
the focus of this pr should be to maintain the uniqueness of the email. even if it's just a sting without a @ sign, let it be. we will make sure the same input doesn't save more than one time.

we will handle validators in different pr since we need in other places too. yes, i think using validator.js is better thank regex.

ashwaryy replied:
Okay. So should I remove the regex for now?

lugenx replied:
yes, please.

AbhiBajpaishek commented on server/src/models/userModel.js:
```diff
@@ -10,7 +10,22 @@ const userSchema = new Schema(
+      lowercase: true,
+      match: [/(.+)@(.+){2,}\.(.+){2,}/, "Please enter a valid email"],
+      validate: {
+        validator: async function (email) {
+          const user = await this.constructor.findOne({ email });
+          if (user) {
+            if (this.id === user.id) {
+              return true;
+            }
+            return false;
+          }
+          return true;
+        },
+        message: () => "Email address already registered.",
+      },
```
any specific reason we haven't used a unique attribute here? I feel its better to add unique attribute here and if validation errors are all handled at one single place rather than doing it here in the model itself.

ashwaryy replied:
Using `unique` option in mongoose would basically check the email field index for uniqueness. But the shortcoming of this approach is that the index must be created on the database for this to work. If index is not created, unique would not work. This is pointed out [here](https://mongoosejs.com/docs/validation.html#the-unique-option-is-not-a-validator).

lugenx replied:
should we put this check inside the static logging method at the bottom of the file?

nikhil-goyal24 replied:
Hii @lugenx ,why are we adding validation to the userModel route we can simply check this in the userController route. 

![image](https://user-images.githubusercontent.com/106306508/211472984-f3cc6148-3587-4631-9932-ac241750b0a6.png)

I am new to the open source and just learning how to contribute, I hope it helps.

lugenx replied:
yeah, this is another way of doing that, and i think it would work just fine, but i still would prefer to keep user validations in the user model.

ashwaryy replied:
> inside the static method

We can do that. Let me know if you'd like me to make any change other than this and removing the regex.

lugenx commented:
@ashwaryy did you get a chance to look into this?

mannan0819 commented on server/src/models/userModel.js:
```diff
@@ -10,7 +10,22 @@ const userSchema = new Schema(
+      match: [/(.+)@(.+){2,}\.(.+){2,}/, "Please enter a valid email"],
+      validate: {
+        validator: async function (email) {
+          const user = await this.constructor.findOne({ email });
```
We need to account for captials and lowercase. Emails are not case sensitive. 
`test@example.com` and `Test@example.com` would be treated differently here but are the same email address

mannan0819 replied:
Solution I can think of is we always do `.tolowercase()` on any user input. So before saving to db we lowercase it and before comparing for login or whether or not the email already exists we lowercase it again

ashwaryy replied:
@mannan0819  I believe `lowercase: true`, which is already added, should be able to take care of it.

mannan0819 replied:
Sweet! Mongoose validation is sick!

mannan0819 reviewed (changes requested):
We need to consider the case sensitivity of emails.

ashwaryy commented:
> @ashwaryy did you get a chance to look into this?

Hi. I have removed the regex and added the validator function at the bottom of the file. Apart from that, I believe things like trimming the email string and ensuring that it's actually an email must happen in the controller before it is sent to the database. There is, however, an option in mongoose to set `trim` to `true`.

lugenx reviewed (commented):
thank you, @ashwaryy, this works great. **one last thing**, let's add the easy one too, add unique to the email field in the schema. after that, it's good to go. 👍

ashwaryy commented on server/src/models/userModel.js:
```diff
@@ -10,7 +10,12 @@ const userSchema = new Schema(
     email: {
       type: String,
-      required: true,
+      lowercase: true,
```
Done

lugenx reviewed (approved):
Awesome, thank you @ashwaryy. great work 👍
